File: app/controllers/iop/collaboration/impl/CollaborativeExamController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package controllers.iop.collaboration.impl;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import impl.EmailComposer;
import io.ebean.DB;
import java.net.URL;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.inject.Inject;
import models.Exam;
import models.ExamExecutionType;
import models.ExamType;
import models.GradeScale;
import models.Language;
import models.User;
import models.json.CollaborativeExam;
import models.sections.ExamSection;
import org.apache.pekko.actor.ActorSystem;
import org.joda.time.DateTime;
import play.libs.Json;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;
import sanitizers.Attrs;
import sanitizers.EmailSanitizer;
import sanitizers.ExamUpdateSanitizer;
import scala.concurrent.duration.Duration;
import security.Authenticated;
import util.config.ConfigReader;

public class CollaborativeExamController extends CollaborationController {

    @Inject
    private ConfigReader configReader;

    @Inject
    private ActorSystem as;

    @Inject
    private EmailComposer composer;

    private Exam prepareDraft(User user) {
        ExamExecutionType examExecutionType = DB.find(ExamExecutionType.class)
            .where()
            .eq("type", ExamExecutionType.Type.PUBLIC.toString())
            .findOne();
        Exam exam = new Exam();
        exam.generateHash();
        exam.setState(Exam.State.DRAFT);
        exam.setExecutionType(examExecutionType);
        cleanUser(user);
        exam.setCreatorWithDate(user);

        ExamSection examSection = new ExamSection();
        examSection.setCreatorWithDate(user);

        examSection.setId(newId());
        examSection.setExam(exam);
        examSection.setExpanded(true);
        examSection.setSequenceNumber(0);

        exam.getExamSections().add(examSection);
        exam.getExamLanguages().add(DB.find(Language.class, "fi"));
        exam.setExamType(DB.find(ExamType.class, 2)); // Final

        DateTime start = DateTime.now().withTimeAtStartOfDay();
        exam.setPeriodStart(start);
        exam.setPeriodEnd(start.plusDays(1));
        exam.setDuration(configReader.getExamDurations().get(0)); // check
        exam.setGradeScale(DB.find(GradeScale.class).findList().get(0)); // check

        exam.setTrialCount(1);
        exam.setAnonymous(true);

        return exam;
    }

    @Authenticated
    @Restrict({ @Group("ADMIN"), @Group("TEACHER") })
    public CompletionStage<Result> searchExams(Http.Request request, final Optional<String> filter) {
        Optional<URL> url = filter.orElse("").isEmpty() ? parseUrl() : parseUrlWithSearchParam(filter.get(), false);
        if (url.isEmpty()) {
            return wrapAsPromise(internalServerError("i18n_internal_error"));
        }

        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        WSRequest wsRequest = wsClient.url(url.get().toString());
        String homeOrg = configReader.getHomeOrganisationRef();

        Function<WSResponse, Result> onSuccess = response ->
            findExamsToProcess(response)
                .map(items -> {
                    List<JsonNode> exams = items
                        .entrySet()
                        .stream()
                        .map(e -> e.getKey().getExam(e.getValue()))
                        .filter(e -> isAuthorizedToView(e, user, homeOrg))
                        .map(this::serialize)
                        .toList();

                    return ok(Json.newArray().addAll(exams));
                })
                .getOrElseGet(Function.identity());

        return wsRequest.get().thenApplyAsync(onSuccess);
    }

    private CompletionStage<Result> getExam(Long id, Consumer<Exam> postProcessor, User user) {
        String homeOrg = configReader.getHomeOrganisationRef();
        return findCollaborativeExam(id)
            .map(ce ->
                downloadExam(ce).thenApplyAsync(result -> {
                    if (result.isEmpty()) {
                        return notFound("i18n_error_exam_not_found");
                    }
                    Exam exam = result.get();
                    if (!isAuthorizedToView(exam, user, homeOrg)) {
                        return notFound("i18n_error_exam_not_found");
                    }
                    postProcessor.accept(exam);
                    return ok(serialize(exam));
                }))
            .getOrElseGet(Function.identity());
    }

    @Restrict({ @Group("ADMIN"), @Group("TEACHER") })
    public Result listGradeScales() {
        Set<GradeScale> grades = DB.find(GradeScale.class).fetch("grades").where().isNull("externalRef").findSet();
        return ok(grades);
    }

    @Authenticated
    @Restrict({ @Group("ADMIN"), @Group("TEACHER") })
    public CompletionStage<Result> getExam(Long id, Http.Request request) {
        return getExam(id, exam -> {}, request.attrs().get(Attrs.AUTHENTICATED_USER));
    }

    @Authenticated
    @Restrict({ @Group("ADMIN"), @Group("TEACHER") })
    public CompletionStage<Result> getExamPreview(Long id, Http.Request request) {
        return getExam(id, exam -> examUpdater.preparePreview(exam), request.attrs().get(Attrs.AUTHENTICATED_USER));
    }

    @Authenticated
    @Restrict({ @Group("ADMIN") })
    public CompletionStage<Result> createExam(Http.Request request) {
        Optional<URL> url = parseUrl();
        if (url.isEmpty()) {
            return wrapAsPromise(internalServerError());
        }
        WSRequest wsRequest = wsClient.url(url.get().toString());
        Function<WSResponse, Result> onSuccess = response -> {
            JsonNode root = response.asJson();
            if (response.getStatus() != CREATED) {
                return internalServerError(root.get("message").asText("Connection refused"));
            }
            String externalRef = root.get("id").asText();
            String revision = root.get("rev").asText();
            CollaborativeExam ce = new CollaborativeExam();
            ce.setExternalRef(externalRef);
            ce.setRevision(revision);
            ce.setCreated(DateTime.now());
            ce.setAnonymous(true);
            ce.save();
            return created(Json.newObject().put("id", ce.getId()));
        };
        Exam body = prepareDraft(request.attrs().get(Attrs.AUTHENTICATED_USER));
        return wsRequest.post(serialize(body)).thenApplyAsync(onSuccess);
    }

    @Restrict({ @Group("ADMIN") })
    public CompletionStage<Result> deleteExam(Long id) {
        return findCollaborativeExam(id)
            .map(ce -> {
                if (!ce.getState().equals(Exam.State.DRAFT) && !ce.getState().equals(Exam.State.PRE_PUBLISHED)) {
                    return wrapAsPromise(forbidden("i18n_exam_removal_not_possible"));
                }
                return examLoader
                    .deleteExam(ce)
                    .thenApplyAsync(result -> {
                        if (result.status() == OK) {
                            ce.delete();
                        }
                        return result;
                    });
            })
            .getOrElseGet(Function.identity());
    }

    @Authenticated
    @With(ExamUpdateSanitizer.class)
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public CompletionStage<Result> updateExam(Long id, Http.Request request) {
        String homeOrg = configReader.getHomeOrganisationRef();
        return findCollaborativeExam(id)
            .map(ce -> {
                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                return downloadExam(ce).thenComposeAsync(result -> {
                    if (result.isPresent()) {
                        Exam exam = result.get();
                        if (isAuthorizedToView(exam, user, homeOrg)) {
                            Exam.State previousState = exam.getState();
                            Optional<Result> error = Stream.of(
                                examUpdater.updateTemporalFieldsAndValidate(exam, user, request),
                                examUpdater.updateStateAndValidate(exam, user, request)
                            )
                                .filter(Optional::isPresent)
                                .map(Optional::get)
                                .findFirst();
                            if (error.isPresent()) {
                                return wrapAsPromise(error.get());
                            }
                            Exam.State nextState = exam.getState();
                            boolean isPrePublication =
                                previousState != Exam.State.PRE_PUBLISHED && nextState == Exam.State.PRE_PUBLISHED;
                            examUpdater.update(exam, request, user.getLoginRole());
                            return uploadExam(ce, exam, user).thenApplyAsync(result2 -> {
                                if (result2.status() == OK && isPrePublication) {
                                    Set<String> receivers = exam
                                        .getExamOwners()
                                        .stream()
                                        .map(User::getEmail)
                                        .collect(Collectors.toSet());
                                    as
                                        .scheduler()
                                        .scheduleOnce(
                                            Duration.create(1, TimeUnit.SECONDS),
                                            () -> composer.composeCollaborativeExamAnnouncement(receivers, user, exam),
                                            as.dispatcher()
                                        );
                                }
                                return result2;
                            });
                        }
                        return wrapAsPromise(forbidden("i18n_error_access_forbidden"));
                    }
                    return wrapAsPromise(notFound());
                });
            })
            .getOrElseGet(Function.identity());
    }

    @Authenticated
    @Restrict({ @Group("ADMIN") })
    public CompletionStage<Result> updateLanguage(Long id, String code, Http.Request request) {
        return findCollaborativeExam(id)
            .map(ce -> {
                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                return downloadExam(ce).thenComposeAsync(result -> {
                    if (result.isPresent()) {
                        Exam exam = result.get();
                        Optional<Result> error = examUpdater.updateLanguage(exam, code, user);
                        return error.isPresent() ? wrapAsPromise(error.get()) : uploadExam(ce, exam, user);
                    }
                    return wrapAsPromise(notFound());
                });
            })
            .getOrElseGet(Function.identity());
    }

    @Authenticated
    @With(EmailSanitizer.class)
    @Restrict({ @Group("ADMIN") })
    public CompletionStage<Result> addOwner(Long id, Http.Request request) {
        return findCollaborativeExam(id)
            .map(ce -> {
                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                return downloadExam(ce).thenComposeAsync(result -> {
                    if (result.isPresent()) {
                        Exam exam = result.get();
                        User owner = createOwner(request.attrs().get(Attrs.EMAIL));
                        exam.getExamOwners().add(owner);
                        return uploadExam(ce, exam, user, owner, null);
                    }
                    return wrapAsPromise(notFound());
                });
            })
            .getOrElseGet(Function.identity());
    }

    @Authenticated
    @Restrict({ @Group("ADMIN") })
    public CompletionStage<Result> removeOwner(Long id, Long oid, Http.Request request) {
        return findCollaborativeExam(id)
            .map(ce -> {
                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                return downloadExam(ce).thenComposeAsync(result -> {
                    if (result.isPresent()) {
                        Exam exam = result.get();
                        User owner = new User();
                        owner.setId(oid);
                        exam.getExamOwners().remove(owner);
                        return uploadExam(ce, exam, user);
                    }
                    return wrapAsPromise(notFound());
                });
            })
            .getOrElseGet(Function.identity());
    }

    private User createOwner(String email) {
        User user = new User();
        user.setId(newId());
        user.setEmail(email);
        return user;
    }
}

File: app/features/iop/collaboration/controllers/CollaborativeExamController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package features.iop.collaboration.controllers

import database.{EbeanJsonExtensions, EbeanQueryExtensions}
import features.iop.collaboration.services.*
import io.ebean.DB
import models.exam.*
import models.sections.ExamSection
import models.user.{Language, Role, User}
import org.joda.time.DateTime
import play.api.Logging
import play.api.data.validation.{Constraints, Invalid}
import play.api.libs.json.*
import play.api.libs.json.Json.*
import play.api.libs.ws.JsonBodyWritables
import play.api.mvc.*
import security.Auth.{AuthenticatedAction, authorized, subjectNotPresent}
import security.{Auth, BlockingIOExecutionContext}
import services.config.ConfigReader
import services.exam.ExamUpdater
import services.mail.EmailComposer
import system.AuditedAction
import validation.exam.ExamValidator

import javax.inject.Inject
import scala.concurrent.Future
import scala.concurrent.duration.*
import scala.jdk.CollectionConverters.*

class CollaborativeExamController @Inject() (
    wsClient: play.api.libs.ws.WSClient,
    examUpdater: ExamUpdater,
    examLoader: CollaborativeExamLoaderService,
    configReader: ConfigReader,
    emailComposer: EmailComposer,
    collaborativeExamService: CollaborativeExamService,
    collaborativeExamSearchService: CollaborativeExamSearchService,
    collaborativeExamAuthorizationService: CollaborativeExamAuthorizationService,
    authenticated: AuthenticatedAction,
    audited: AuditedAction,
    override val controllerComponents: ControllerComponents
)(implicit ec: BlockingIOExecutionContext)
    extends BaseController
    with EbeanQueryExtensions
    with EbeanJsonExtensions
    with JsonBodyWritables
    with Logging:

  private def prepareDraft(user: User): Exam =
    val examExecutionType = DB
      .find(classOf[ExamExecutionType])
      .where()
      .eq("type", ExamExecutionType.Type.PUBLIC.toString)
      .findOne()

    val exam = new Exam()
    exam.generateHash()
    exam.setState(Exam.State.DRAFT)
    exam.setExecutionType(examExecutionType)
    CollaborativeExamProcessingService.cleanUser(user)
    exam.setCreatorWithDate(user)

    val examSection = new ExamSection()
    examSection.setCreatorWithDate(user)
    examSection.setId(CollaborativeExamProcessingService.newId())
    examSection.setExam(exam)
    examSection.setExpanded(true)
    examSection.setSequenceNumber(0)

    exam.getExamSections.add(examSection)

    exam.getExamLanguages.add(DB.find(classOf[Language], "fi"))
    exam.setExamType(DB.find(classOf[ExamType], 2)) // Final
    exam.setGradeScale(DB.find(classOf[GradeScale]).list.head)

    val start = DateTime.now().withTimeAtStartOfDay()
    exam.setPeriodStart(start)
    exam.setPeriodEnd(start.plusDays(1))
    exam.setDuration(configReader.getExamDurationsJava.asScala.head)

    exam.setTrialCount(1)
    exam.setAnonymous(true)
    exam.setGradingType(Grade.Type.GRADED)

    exam

  def searchExams(filter: Option[String]): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.TEACHER))).async { request =>
      val user      = request.attrs(Auth.ATTR_USER)
      val homeOrg   = configReader.getHomeOrganisationRef
      val wsRequest = collaborativeExamSearchService.buildSearchRequest(filter)

      wsRequest.get().flatMap { response =>
        collaborativeExamSearchService.findExamsToProcess(response).map {
          case Left(result) => result
          case Right(items) =>
            val exams = items
              .filter { case (ce, rev) =>
                // Deserialize to check authorization
                val exam = ce.getExam(toJacksonJson(rev))
                collaborativeExamAuthorizationService.isAuthorizedToView(exam, user, homeOrg)
              }
              .map { case (ce, rev) =>
                // Add local database ID to the JSON response
                // The external service uses _id (CouchDB style), but we need the local database id
                val jsonObj = rev.as[JsObject]
                // Ensure the local database id is set (overwrite any existing id field)
                jsonObj + ("id" -> JsNumber(BigDecimal(ce.getId)))
              }
              .toSeq

            Ok(JsArray(exams))
        }
      }
    }

  private def getExam(id: Long, postProcessor: Exam => Unit, user: User): Future[Result] =
    val homeOrg = configReader.getHomeOrganisationRef
    collaborativeExamAuthorizationService.findCollaborativeExam(id).flatMap {
      case Left(errorResult) => Future.successful(errorResult)
      case Right(ce) =>
        examLoader.downloadExamJson(ce).map {
          case None       => NotFound("i18n_error_exam_not_found")
          case Some(root) =>
            // Deserialize to check authorization
            val jacksonNode = play.libs.Json.parse(Json.stringify(root))
            val exam        = ce.getExam(jacksonNode)
            if !collaborativeExamAuthorizationService.isAuthorizedToView(exam, user, homeOrg) then
              NotFound("i18n_error_exam_not_found")
            else
              postProcessor(exam)
              // Add local database ID to the JSON response
              val jsonObj    = root.as[JsObject]
              val jsonWithId = jsonObj + ("id" -> JsNumber(BigDecimal(ce.getId)))
              Ok(jsonWithId)
        }
    }

  def listGradeScales(): Action[AnyContent] =
    controllerComponents.actionBuilder.andThen(subjectNotPresent) { _ =>
      val grades = DB
        .find(classOf[GradeScale])
        .fetch("grades")
        .where()
        .isNull("externalRef")
        .distinct
      Ok(grades.asJson)
    }

  def getExam(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.TEACHER))).async { request =>
      getExam(id, _ => (), request.attrs(Auth.ATTR_USER))
    }

  def getExamPreview(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.TEACHER))).async { request =>
      getExam(id, exam => examUpdater.preparePreview(exam), request.attrs(Auth.ATTR_USER))
    }

  def createExam(): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.ADMIN))).async { request =>
      collaborativeExamSearchService.parseExamsUrl() match
        case None => Future.successful(InternalServerError)
        case Some(url) =>
          val user      = request.attrs(Auth.ATTR_USER)
          val body      = prepareDraft(user)
          val wsRequest = wsClient.url(url.toString)

          wsRequest.post(body.asJson).flatMap { response =>
            if response.status != CREATED then
              Future.successful(
                InternalServerError(
                  (response.json \ "message").asOpt[String].getOrElse("Connection refused")
                )
              )
            else
              val externalRef = (response.json \ "id").as[String]
              val revision    = (response.json \ "rev").as[String]
              collaborativeExamService.create(externalRef, revision, anonymous = true).map { ce =>
                Created(Json.obj("id" -> JsNumber(BigDecimal(ce.getId))))
              }
          }
    }

  def deleteExam(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))).async { _ =>
      collaborativeExamAuthorizationService.findCollaborativeExam(id).flatMap {
        case Left(errorResult) => Future.successful(errorResult)
        case Right(ce)
            if ce.getState != Exam.State.DRAFT && ce.getState != Exam.State.PRE_PUBLISHED =>
          Future.successful(Forbidden("i18n_exam_removal_not_possible"))
        case Right(ce) =>
          examLoader.deleteExam(ce).flatMap { result =>
            if result.header.status == Ok.header.status then
              collaborativeExamService.delete(ce).map(_ => result)
            else Future.successful(result)
          }
      }
    }

  def updateExam(id: Long): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN)))
    .async(controllerComponents.parsers.json) { request =>
      // Validate the exam payload
      ExamValidator.forUpdate(request.body) match
        case Left(ex) => Future.successful(BadRequest(ex.getMessage))
        case Right(payload) =>
          val homeOrg = configReader.getHomeOrganisationRef
          collaborativeExamAuthorizationService.findCollaborativeExam(id).flatMap {
            case Left(errorResult) => Future.successful(errorResult)
            case Right(ce) =>
              val user = request.attrs(Auth.ATTR_USER)

              examLoader.downloadExam(ce).flatMap {
                case None => Future.successful(NotFound("i18n_error_exam_not_found"))
                case Some(exam)
                    if !collaborativeExamAuthorizationService.isAuthorizedToView(
                      exam,
                      user,
                      homeOrg
                    ) =>
                  Future.successful(Forbidden("i18n_error_access_forbidden"))
                case Some(exam) =>
                  val previousState = exam.getState

                  // Validate temporal fields and state
                  val error = Seq(
                    examUpdater.updateTemporalFieldsAndValidate(exam, user, payload),
                    examUpdater.updateStateAndValidate(exam, user, payload)
                  ).flatten.headOption

                  error match
                    case Some(err) => Future.successful(err)
                    case None =>
                      val nextState = exam.getState
                      val isPrePublication =
                        previousState != Exam.State.PRE_PUBLISHED && nextState == Exam.State.PRE_PUBLISHED

                      examUpdater.update(exam, payload, user.getLoginRole)

                      examLoader.uploadExam(ce, exam, user).map { result =>
                        if result.header.status == Ok.header.status && isPrePublication then
                          val receivers = exam.getExamOwners.asScala.map(_.getEmail).toSet
                          emailComposer.scheduleEmail(1.second) {
                            emailComposer.composeCollaborativeExamAnnouncement(
                              receivers,
                              user,
                              exam
                            )
                          }
                        result
                      }
              }
          }
    }

  def updateLanguage(id: Long, code: String): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.ADMIN))).async { request =>
      collaborativeExamAuthorizationService.findCollaborativeExam(id).flatMap {
        case Left(errorResult) => Future.successful(errorResult)
        case Right(ce) =>
          val user = request.attrs(Auth.ATTR_USER)
          examLoader.downloadExam(ce).flatMap {
            case None => Future.successful(NotFound("i18n_error_exam_not_found"))
            case Some(exam) =>
              examUpdater.updateLanguage(exam, code, user) match
                case Some(error) => Future.successful(error)
                case None        => examLoader.uploadExam(ce, exam, user)
          }
      }
    }

  def addOwner(id: Long): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.ADMIN)))
    .async(controllerComponents.parsers.json) { request =>
      collaborativeExamAuthorizationService.findCollaborativeExam(id).flatMap {
        case Left(errorResult) => Future.successful(errorResult)
        case Right(ce) =>
          val user = request.attrs(Auth.ATTR_USER)
          (request.body \ "email").asOpt[String] match
            case None        => Future.successful(BadRequest("i18n_error_email_missing"))
            case Some(email) =>
              // Validate email
              Constraints.emailAddress()(email) match
                case Invalid(_) => Future.successful(BadRequest("i18n_error_email_invalid"))
                case _ =>
                  examLoader.downloadExam(ce).flatMap {
                    case None => Future.successful(NotFound("i18n_error_exam_not_found"))
                    case Some(exam) =>
                      val owner = createOwner(email)
                      exam.getExamOwners.add(owner)
                      examLoader.uploadExam(ce, exam, user, owner, null)
                  }
      }
    }

  def removeOwner(id: Long, oid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))).async { request =>
      collaborativeExamAuthorizationService.findCollaborativeExam(id).flatMap {
        case Left(errorResult) => Future.successful(errorResult)
        case Right(ce) =>
          val user = request.attrs(Auth.ATTR_USER)
          examLoader.downloadExam(ce).flatMap {
            case None => Future.successful(NotFound("i18n_error_exam_not_found"))
            case Some(exam) =>
              val owner = new User()
              owner.setId(oid)
              exam.getExamOwners.remove(owner)
              examLoader.uploadExam(ce, exam, user)
          }
      }
    }

  private def createOwner(email: String): User =
    val user = new User()
    user.setId(CollaborativeExamProcessingService.newId())
    user.setEmail(email)
    user
